### ISSUE TITLE
Key repeat lags over SSH connection (only)

### ISSUE THREAD

**Issue by fraimondo** [nushell/nushell#14474]
### Describe the bug

I have installed nushell locally and in a remote server.

When I use it locally, the shell is working perfectly. Both zsh and nu are blazinigly fast when I want to move left/right through a long command or when I search the command history with up/down.

When I connect remotely, zsh is fast, but nu has a considerable delay/lag that makes it unusable.

I tried with ITerm2, Alacritty, MacOS Terminal and Kitty.

Though with ITerm2 I have the same slow behaviour locally, with the other 3 terminal apps I only have the issue with nu on a remote location.

I tried changing the `shell_integration` config section, but that did not seem to change the behaviour.

### How to reproduce

1. open zsh locally, type a long command. Keep the left key pressed
2. do the same with nu, locally
3. do the same with zsh, on a remote ssh machine
4. do the same with nu, on a remote ssh machine

### Expected behavior

I expected nu to work equally fast in both locations.

### Configuration

| key                | value                               |
| ------------------ | ----------------------------------- |
| version            | 0.99.1                              |
| major              | 0                                   |
| minor              | 99                                  |
| patch              | 1                                   |
| branch             |                                     |
| commit_hash        |                                     |
| build_os           | linux-x86_64                        |
| build_target       | x86_64-unknown-linux-gnu            |
| rust_version       | rustc 1.82.0 (f6e511eec 2024-10-15) |
| rust_channel       | stable-x86_64-unknown-linux-gnu     |
| cargo_version      | cargo 1.82.0 (8f40fc59f 2024-08-21) |
| build_time         | 2024-11-12 13:54:21 +01:00          |
| build_rust_channel | release                             |
| allocator          | mimalloc                            |
| features           | default, sqlite, trash              |
| installed_plugins  |                                     |

**Comment by fdncred** [comment]
Do you experience the same behavior in 0.100.0?

**Comment by fraimondo** [comment]
0.100.0 is even worse.

Is there any way to get a debug log?

**Comment by fdncred** [comment]
maybe nu --log-level trace

**Comment by fraimondo** [comment]
So after debugging and checking the trace level, it looks like it's not an input lag, but rather a rendering/output one, strictly related to the latency to the SSH server.

With a ping of ~1ms (at work), nushell behaves exactly like zsh, responsive and fast.

With a ping of ~50ms (outside work), nushell is almost unusable, zsh is still responsive and fast.

Is there any configuration setting I can change to see what is causing this?

**Comment by fdncred** [comment]
I wonder if it's something in reedline or if it's the implementation of it in nushell? Is it possible to run any of the reedline examples over SSH to see? This would at least narrow down where to look.

**Comment by fraimondo** [comment]
So I ran the `basic` reedline example, wrote `cd ` and a very long path and then pressed the right/left keys. The lag is there.

Same test on the zsh shell I used to run the example was fast.

**Comment by fdncred** [comment]
ok, at least we've narrowed it down between the large codebase of nushell and smaller codebase of reedline. Now someone needs to try and figure out what's going on. Seems like there was some type of timer around keypresses, iirc, but I don't remember why. Maybe for detecting pastes or something? 🤔

**Comment by DavidTruby** [comment]
As an extra data point, this doesn't occur when ssh-ing to a Windows machine. It seems it only happens on linux for some reason.

**Comment by mxk** [comment]
Seeing the same issue with PuTTY and Windows OpenSSH when connecting to Ubuntu 20.04. I couldn't use the pre-built gnu version because my glibc was too old. Tried musl and observed this behavior. Thought it was related to that, so built with cargo, but no difference.

**Comment by seckinu** [comment]
Seeing the same issue, connecting to a Ubuntu 24.02 machine through Windows Terminal. Zsh responsive, Nu annoyingly laggy.

**Comment by sam701** [comment]
Any progress on this topic? This is the only issue preventing me from using nushell on remote servers. I have to fall back to using fish.

**Comment by fdncred** [comment]
What this takes is someone familiar with ssh and nushell to see what's going and then submit a PR to reedline or nushell or both to fix it. We're still waiting for such a contributor to reveal themselves.

**Comment by zanexGHG** [comment]
Over the past couple months since I also had this problem I made a little wrapper for nushell which does the job (the bash_history only works for the current session yet) and I think its pretty helpful if anyone needs it:

```bash
function nu_wrap() {
    while true; do
        USER="$(whoami)"
        HOST="example.com"
        CUR_DIR="$(pwd | sed "s|$HOME|~|")"

        echo -e "┌ \033[01;31m${USER}\033[01;33m@\033[01;36m${HOST} \033[01;33m${CUR_DIR}\033[0m"
        read -er -p "└$ " cmd

        [[ -z "$cmd" ]] && continue

        history -s "$cmd"

        if [[ "$cmd" == "exit" ]]; then
            break
        fi

        if [[ "$cmd" == cd* ]]; then
            dir=$(echo "$cmd" | cut -d' ' -f2-)
            if [ -z "$dir" ]; then
                dir="$HOME"
            fi
            cd "$dir" 2>/dev/null || echo "No such directory: $dir"
        else
            nu -c "$cmd"
        fi
    done
}
if [[ $- == *i* ]]; then
    nu_wrap
fi
```

**Comment by haras-unicorn** [comment]
does anyone else experience this being even worse when running `journalctl` or `systemctl status` ?

**Comment by sam701** [comment]
> does anyone else experience this being even worse when running `journalctl` or `systemctl status` ?

I didn't notice a difference with journalctl but I do see a huge improvement when using nu within tmux or zellij.

**Comment by wonrax** [comment]
It's also way more responsive inside tmux for me.

**Comment by k3-cat** [comment]
Hi everyone, I am so surprised that there actually are people who discusses this issue. Since I found nothing when I encounter the same problem when using pwsh. I did some very preliminary tests, and hope them can be helpful.

#### My env:
```
Host OS:  Windows (26220.6690)
Host SSH: OpenSSH_for_Windows_9.5p1, LibreSSL 3.8.2
Server OS:  Arch
Server SSH: this issue lasts for about / more than one year, so versions doesn't matter
* The server does not have any GPUs
** Tried `ssh -t` but no noticeable changes 

Latency: generally, > 100ms

Terminal (Host): Windows Terminal
Tested Shells:   nushell, pwsh (yes, exactly the same behaviour)
```

#### My findings:
It gets resolved if shells are running within something like `screen`. However, `su` or `sudo / doas / run0 -u <user> nu` doesn't -- if something can break `Starship` (which doesn't support 24-bits colours and only display a white text), it can resolve this problem.

**Comment by HammerPoppy** [comment]
making nushell unusable for me unfortunately

**Comment by fdncred** [comment]
I thought this was fixed in the latest main branch? Anyone tested?

**Comment by kronberger-droid** [comment]
Well I would say fixed is a very relative term here. I think we improved it noticeably with https://github.com/nushell/reedline/pull/1090, maybe @reubeno has some metrics on how much it changed.
But there are probably even more performance improvements we can do.

**Comment by reubeno** [comment]
> Well I would say fixed is a very relative term here. I think we improved it noticeably with https://github.com/nushell/reedline/pull/1090, maybe @reubeno has some metrics on how much it changed.
> But there are probably even more performance improvements we can do.
> 

Agreed that there are more opportunities for improvement. The linked PR was a first step.

I didn't specifically measure key-repeat latency, but if users are seeing significantly different performance between shell-over-ssh vs. shell-in-tmux-over-ssh, then that does sound like what the linked PR started addressing. (It implies delay from round trip with the remote terminal emulator.)
